docs/src/lang/variants.md

```diff
@@ -0,0 +1,351 @@
+# Variants
+
+**Warning:** This feature is currently under testing. To enable variants
+support in Apalache, pass the option `--features=rows`.
+
+[[Back to all operators]](./apalache-extensions.md)
+
+[Variants][] (also called *tagged unions*) are useful, when you want to combine
+values of different shapes in a single set or a sequence.
+
+**Idiomatic tagged unions in untyped TLA+.** In untyped TLA+, one can simply
+pack a value into a record and add an additional tag field. For instance, we
+could create a set that contains two records of different shapes:
+
+```tla
+ApplesAndOranges == {
+    [ tag |-> "Apple", color |-> "red" ],
+    [ tag |-> "Orange", seedless |-> TRUE ]
+  }
+```
+
+Now we can unpack the elements of `ApplesAndOranges` based on their tag:
+
+```tla
+  \E e \in ApplesAndOranges:
+    /\ e.tag = "Apple"
+    /\ e.color /= "green"
+```
+
+This idiom is quite common in untyped TLA+. [Tagged unions in Paxos][] is
+probably the most illuminating example of this idiom. Unfortunately, it is way
+too easy to make a typo in the tag name, since it is a string, or simply access
+a wrong record without checking its tag.
+
+**Variants module.** Apalache formalizes the above idiom in the module
+[Variants.tla][]. Due to the type checker, the users get a type error, when
+they access a wrong value. Additionally, the default implementation raises an
+error in TLC, when a variant is used incorrectly.
+
+**Immutability**. As variants are immutable.
+
+**Construction.** An instance of a variant can be constructed via the operator
+`Variant`:
+
+```tla
+Variant("Apple", "red")
+```
+
+If we just construct a variant like in the example above, it will be assigned
+a parametric variant type:
+
+```
+Apple(Str) | a
+```
+
+In this type, we know that whenever a value is tagged with "Apple" it should be
+of the string type. However, we know nothing about other options. Most of the
+time, we want to define variants that are sealed, that is, we know all
+available options. In the Apalache terms, it means that our type would look
+like follows:
+
+```
+Apple(Str) | Orange(Bool)
+```
+
+To do this, we can introduce variants together with user-defined
+constructors for each option right away:
+
+```tla
+\* @typeAlias: FRUIT = Apple(Str) | Orange(Bool);
+\*
+\* @type: Str => FRUIT;
+Apple(color) == Variant("Apple", color)
+
+\* @type: Bool => FRUIT;
+Orange(seedless) == Variant("Orange", seedless)
```

Comment:
Nitpicking here, but do you consider this interleaved `\*` idiomatic?
I've seen it used below as well, but I find the grouping confusing, because then `Orange` is offset by an empty line.

```suggestion
\* @typeAlias: FRUIT = Apple(Str) | Orange(Bool);

\* @type: Str => FRUIT;
Apple(color) == Variant("Apple", color)

\* @type: Bool => FRUIT;
Orange(seedless) == Variant("Orange", seedless)
```

Reply:
Makes sense, though we have to make this works with SANY.